core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -789,76 +865,45 @@ private int getTotalDynamicFilters()
+                verify(collectionContext != null, "collectionContext is missing for %s", dynamicFilterId);
+                collectionContext.collect(taskId, domain);
+            });
 
-        private Map<DynamicFilterId, List<Domain>> getTaskDynamicFilters(StageId stageId, Optional<Set<DynamicFilterId>> selectedFilters)
-        {
-            return selectedFilters.orElseGet(() -> stageDynamicFilters.get(stageId)).stream()
-                    .collect(toImmutableMap(
-                            identity(),
-                            filter -> Optional.ofNullable(taskDynamicFilters.get(filter))
-                                    .map(taskDomains -> ImmutableList.copyOf(taskDomains.values()))
-                                    // return empty list in case filter has already been collected and task domains have been removed
-                                    .orElse(ImmutableList.of())));
+            if (stageDynamicFilters.computeIfAbsent(taskId.getStageId(), key -> newConcurrentHashSet()).addAll(newDynamicFilters.keySet())) {
+                updateExpectedTaskCount();
+            }
         }
 
-        private void addDynamicFilters(Map<DynamicFilterId, List<Domain>> newDynamicFilters)
+        private void stageCannotScheduleMoreTasks(StageId stageId, int numberOfTasks)
         {
-            SetMultimap<Consumer<Map<DynamicFilterId, Domain>>, DynamicFilterId> completedConsumers = HashMultimap.create();
-            newDynamicFilters.forEach((filter, domain) -> {
-                if (taskDynamicFilters.remove(filter) == null) {
-                    // filter has been collected concurrently
-                    return;
-                }
-                dynamicFilterCollectionTime.put(filter, System.nanoTime());
-                dynamicFilterSummaries.put(filter, union(domain));
-                Optional.ofNullable(lazyDynamicFilters.get(filter)).ifPresent(future -> future.set(null));
-                List<Consumer<Map<DynamicFilterId, Domain>>> consumers;
-                synchronized (dynamicFilterConsumers) {
-                    // this section is executed only once due to the earlier null check on taskDynamicFilters.remove(filter)
-                    consumers = requireNonNull(dynamicFilterConsumers.remove(filter));
-                }
-                // dynamic filter updates are batched up per-consumer to reduce number of callbacks
-                consumers.forEach(consumer -> completedConsumers.put(consumer, filter));
-            });
-            completedConsumers.asMap().forEach((consumer, dynamicFilterIds) -> consumer.accept(
-                    dynamicFilterIds.stream()
-                            .collect(toImmutableMap(
-                                    identity(),
-                                    filterId -> requireNonNull(dynamicFilterSummaries.get(filterId))))));
+            if (stageNumberOfTasks.put(stageId, numberOfTasks) == null) {
```

Comment:
Should this be `checkState` or it's valid for `stageCannotScheduleMoreTasks` to get called multiple times for a StageId ?

Reply:
With the current implementation it is possible for this method to be called more than once. I'm not sure if that's the right approach though. I guess we can change it later as a separate refactor.